### ISSUE TITLE
Request: An option to show "undefined" value as a string (JSON mode)

### ISSUE THREAD

**Issue by MLefebvreICO** [mradionov/karma-jasmine-diff-reporter#17]
Hi there,

Would it be possible that you add an option that we could set to `true` to allow `undefined` values in the JSON output.

By default, `JSON.stringify` removes all `undefined` value, so we would have to replace it with a `undefined` string.

So let's say we have the following : 

``` js
var obj1 = {Foo: 'Foo', Bar: undefined};
var obj2 = {Foo: 'Foo'};

console.log(JSON.stringify(obj1));
// '{"Foo": "Foo"}'
console.log(JSON.stringify(obj2));
// '{"Foo": "Foo"}'
```

So it would do this while testing, but we cannot see why it is not passing : 

``` json
Expected '{"Foo": "Foo"}' to equal '{"Foo": "Foo"}'
```

**Comment by mradionov** [comment]
Seems like it really is an issue, because it was requested before in the issue https://github.com/mradionov/karma-jasmine-diff-reporter/issues/14 . My answer there was:

> I don't think it's a bug, undefined is not a supported value for JSON format. I would like to keep output compatible with JSON specs without bringing in any custom mixins.

Let me think once again about it, maybe I'll come up with something. Thanks!

**Comment by mradionov** [comment]
Could you please clarify to me one thing: do you use reporter's `json` option set to `true` to get JSON output?

**Comment by BleuShan** [comment]
> Could you please clarify to me one thing: do you use reporter's json option set to true to get JSON output?

@mradionov Yes. Come to think of it, IIRC, function members are also elided, which can also cause some problems.

**Comment by mradionov** [comment]
And there are also other constructs, which get removed when passed to `JSON.stringify`, I've mentioned it in the issue, where the feature was added - https://github.com/mradionov/karma-jasmine-diff-reporter/issues/14#issuecomment-227874725 . The problem is that when I start adding customizations on top of JSON output (for example keep `undefined` or keep function members), then:

1) it won't longer be a valid JSON
2) if ignore first point, by adding all these missing constructs back, I'll eventually end up with an output somewhat similar to already existing original Jasmine output

What could be done, is not to keep the output JSON compatible, but keep it JavaScript compatible, so the output could be copied/pasted and used in JS environment. In this case `undefined` and probably other constructs will be a valid part of the output, as we are aiming for valid JS.

**Comment by BleuShan** [comment]
Correct me if i'm wrong, it's valid json if the output looks like:
```json
{
  "FunctionMember":  "function FunctionMember",
  "UndefinedMember: "undefined"
}
```
As it doesn't introduce entirely new primitives and the members will get deserialize as strings by most parsers.

>  [ … ], but keep it JavaScript compatible, so the output could be copied/pasted and used in JS environment.

If you're thinking that it could output JavaScript source code, it's seems a bit of headache because: 

* JavaScript function decompilation is not standardized across JavaScript interpreters. 
* It seems that only Gecko implements `Object.prototype.toSource()`
* It's probably way much more information that we actually need displayed.

The whole point of this issue is that test results can be at times surprising, if you're not used to it.

**Comment by mradionov** [comment]
I fully agree that missing `undefined` prop is an unexpected behavior for test reporters and I take it as a bug. 
The original idea of the feature - to be able to copy/paste output to JS env as a fully functional construct,  **but** it requires to get rid of some complex unserializable constructs like functions as you've mentioned or it will require to format them in some weird way, maybe in a string `"function FunctionMember"`, so output JS will still work, but partially. And what's the point of having partially working construct? Simple literals and arrays/objects with literals will be fine, I could even make `undefined`work, but anything further will still be unavailable.

To be honest I think it was my mistake to bring this feature in, I haven't thought about it that deep. The whole point of a test reporter to be as verbose as possible, my reporter was supposed to increase that verbose-ability a bit by simply highlighting the diff. But I went a little to far trying to make the reporter a tool that solves all the problems. This `json` feature looks a bit out of place for me now, I am not sure what to do with it now: if I keep it, then the bugs with missing props and displaying of complex structures will come up sooner or later; if I remove it, then a reporter will come back to solving it's original problem - highlighting diffs - without going out of boundaries with these extra features, but some users might get angry that the feature is gone :-)

It is interesting though, when one of your projects becomes kinda popular and you get first feedback from your users, about what they don't like and what they would like you to implement, and you immediately go "sure I'll do it, thank you", because nobody was ever paying attention to your projects before, but all of a sudden you have somebody coming and asking to improve you project. It becomes very exciting to have this feedback, and you do things that you won't usually do by simply saying "no".

I did not have to write these last two paragraphs, I tried to explain it to myself and remember this situation :-) Any advice about the feature or in general will be greatly appreciated, I need to think about the feature once again with a clear head.

**Comment by BleuShan** [comment]
> * It's probably way much more information that we actually need displayed.

I was referring to [the output result of a decompiled function](http://kangax.github.io/jstests/function-decompilation/). Given that it is usually the source code of the function and that we use this reporter because it enables us to read tests results faster.

**Comment by MLefebvreICO** [comment]
> Any advice about the feature or in general will be greatly appreciated

See [this](http://stackoverflow.com/questions/26540706/json-stringify-removes-hash-keys-with-undefined-values). You could, instead of putting `null`, putting a string that says `undefined`

**Comment by mradionov** [comment]
@BleuShan , I agree that displaying function source code is not the best idea. Also it is not the best way of comparing functions, because Jasmine uses === to compare them, and even if the string representation is the same, instances might be different. I have to mention that I don't have access to Jasmine matchers code from inside the reporter, because of that I don't know whether functions are different or not. The only info I have is Jasmine output error message. And it is a problem, because even Jasmine simply displays `Function` in a message, so the diff becomes useless:

![image](https://cloud.githubusercontent.com/assets/2007370/17494554/e36a75a2-5dbd-11e6-81ec-53d8db5493fa.png)

I think it's better to extract this functions problem into a separate issue and to think about is separately.

-- 

@MLefebvreICO , to be honest I don't like the idea of displaying `undefined` as a string, because it might confuse a developer that he has a string `"undefined"` except of actual `undefined` in his source code. Displaying real `undefined` is more user-friendly, but again it will break JSON, yet it won't break JS. I am just curious, do you use JSON diff services to compare the results? Ideally the reporter should prevent this kind of copy/paste to external services for comparing strings, as it should show the difference right in the terminal. I am still thinking about what to do with it.

**Comment by BleuShan** [comment]
> I am just curious, do you use JSON diff services to compare the results?

Sometimes when the reporter trip on itself and doesn't display any diff or when we use a custom matcher that we haven't bothered to register to the reporter.

> Displaying real undefined is more user-friendly, but again it will break JSON, yet it won't break JS.

One way to preserve the JSON validity is to have a metadata key appended to the object similar to how graphql outputs its introspection data:

```json
{
   "__JasmineDiffReporterIntrospectionMetadata": {
       "undefinedObjectKeys": [
            "UndefinedMember"
       ],
       "methodObjectKeys": [
            "FunctionMember"
       ]
  }}
```

It's less intuitive (and slower to read in some cases) to read, but it gets the job done?

**Comment by mradionov** [comment]
I am also worried about nested objects in this case.

**Comment by mradionov** [comment]
What I am thinking about now is to make the feature a bit more configurable, so user would have some more options (not solid proof though), which are better then nothing, especially in this kind of situation.

- Allow to display `undefined` value as a string `"undefined"` in JSON output. Originally `undefined` value is completely removed from JSON output.

    ```
    json: {
      undefined: 'string'
    }
    ```

    Example:

    ```
    {
      "foo": "undefined"
    }
    ```

- Allow to display function body in JSON output, because sometimes it might help. Function will be converted to a string using `func.toString()`

    ```
    json: {
      function: 'body'
    }
    ```

    Example:

    ```
    {
      "foo": "function bar (a, b, c) {
        return 42;
      }"
    }
    ```

- Allow to display only function name an args (not sure about this option), use `func.toString` and regular expressions to extract function head:

    ```
    json: {
      function: 'head'
    }
    ```

    Example:

    ```
    {
      "foo": "function bar (a, b, c) { ... }"
    }
    ```


Default value for all options will be `false`, because:

1) displaying `undefined` as a string is an unexpected behavior, even though completely missing `undefined` is also not very expected
2) displaying function as a string is an unexpected behavior, even though it is completely missing from output
3) displaying function body might be too verbose to make it a default

User has to explicitly turn options on in the reporter options. Will be happy to hear any feedback.

**Comment by BleuShan** [comment]
> Default value for all options will be false, because:

> 1) displaying undefined as a string is an unexpected behavior, even though completely missing undefined is also not very expected
2) displaying function as a string is an unexpected behavior, even though it is completely missing from output
3) displaying function body might be too verbose to make it a default

> User has to explicitly turn options on in the reporter options. Will be happy to hear any feedback

It rather have the wrong output than no output and a bunch of questions.

> I am also worried about nested objects in this case.

What do you mean?

**Comment by MLefebvreICO** [comment]
@mradionov I'm ok with the options, but it is not really oblivious for a new comer, Maybe you could add a warning to the `json` flag and provide information like : 

> "This will do a JSON.stringify. Options normally not serialized like `undefined` values and functions' body are not shown. In that case you may get an `Expect` object that looks the same as the diff."

**Comment by mradionov** [comment]
Thanks. I am on vacation now, I'll review it once again and proceed with implementation in two weeks or so.

**Comment by mradionov** [comment]
Looking at it once again it looks overly complicated, these new options will bring to much complexity. I personally would like to remove json option from the reporter, as it makes very little sense in the reporter. If one wants to copy a value from terminal to use in JS environment, for complex values you will always get incomplete results (objects,functions,etc), because it is impossible to recreate them from a string. If one wants to use diff services, you can use text diff services instead of JSON diff services and get pretty the same result.

One simple replacement for json feature could be removing `Object({})` Jasmine wrapper, which will make objects copy-able in some simple cases. It will be a cosmetic change not influencing the output at all.

Regarding functions not highlighting - @MLefebvreICO created an issue #18 , I'll try to tackle them and similar cases in there.

**Comment by mradionov** [comment]
Following my latest comments in here, I've published a new [1.0.0](https://github.com/mradionov/karma-jasmine-diff-reporter/releases/tag/v1.0.0) version of the reporter with `json` option removed.
New `verbose` option ([docs](https://github.com/mradionov/karma-jasmine-diff-reporter#verbose)) might be a decent replacement for it, depending on how you were using `json` option.
